### PR TITLE
1580609 - Add iOS Dispatchers

### PR DESCRIPTION
**Pull request by travis79**
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/ios/GleanTests/DispatchersTest.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+//  GleanTests
+//
+//  Created by Travis Long on 9/12/19.
+//  Copyright © 2019 Jan-Erik Rediger. All rights reserved.
+//
+
+@testable import Glean
+import XCTest
+
+class DispatchersTest: XCTestCase {
```
Comment (Dexterp37) [inline glean-core/ios/GleanTests/DispatchersTest.swift]: Let's make extract sure this isn't flaky :) Unleash your testing obscure power :-D

File: glean-core/ios/Glean/Dispatchers.swift [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    /// Dispatchers.shared.launch {
+    ///     // Do some work here
+    /// }.completionBlock = {
+    ///     print("Done!")
+    /// }
+    /// ```
+    ///
+    /// You could also use the returned `Operation` to await that specific task like this:
+    /// ```
+    /// let operation = Dispatchers.shared.launch {
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: Wait, just assigning it to a variable automatically awaits the job to be finished?

File: glean-core/ios/Glean/Dispatchers.swift [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                // If we are in testing mode, go ahead and execute the operation and wait until it
+                // is finished before continuing to give us synchronous execution of the task.
+                operation.start()
+                operation.waitUntilFinished()
+            } else {
+                operationQueue.addOperation(operation)
+            }
+        }
+
+        return operation
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: (Not critical to landing this, this is internal API so we control all execution paths)

To answer my own question from above to await something returned from `launch` you need to call `waitUntilFinished`.
But what if I call `start` on the thing that is returned? Is this returned operation a clone of what's in the queue and thus things get executed twice? Or is it the same object and the task in the queue will not be executed again later?

File: glean-core/ios/GleanTests/DispatchersTest.swift (hunk 1 above)
Comment (travis79) [reply glean-core/ios/GleanTests/DispatchersTest.swift]: I'm pretty confident in this one, there are better ways of controlling and awaiting the async bits, but I'll try and put it through as much testing as possible.

File: glean-core/ios/Glean/Dispatchers.swift (hunk 3 above)
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: You shouldn't be able to call `start()` on a `finished` operation, if I understand how the `Operation` class works. It needs to be in the `isReady` state.

Comment (badboy) [reply glean-core/ios/Glean/Dispatchers.swift]: What happens then? An exception?

Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: I will look it up again to be sure, but I think it's a no-op.

**Pull request merged**